kaggle_environments/envs/halite/helpers.py

```diff
@@ -0,0 +1,780 @@
+            Returns the ship with the least halite or None in the case of a tie along with all other ships.
+            """
+            if len(ships) == 1:
+                return ships[0], []
+            ships_by_halite = group_by(ships, lambda ship: ship.halite)
+            smallest_halite = min(ships_by_halite.keys())
+            smallest_ships = ships_by_halite[smallest_halite]
+            if len(smallest_ships) == 1:
+                # There was a winner, return it
+                winner = smallest_ships[0]
+                return winner, [ship for ship in ships if ship != winner]
+            # There was a tie for least halite, all are deleted
+            return None, ships
+
+        # Check for ship to ship collisions
+        ship_collision_groups = group_by(board.ships.values(), lambda ship: ship.position)
+        for position, collided_ships in ship_collision_groups.items():
+            winner, deleted = resolve_collision(collided_ships)
+            if winner is not None:
+                winner.cell._ship_id = winner.id
+            for ship in deleted:
+                board._delete_ship(ship)
+                if winner is not None:
+                    # Winner takes deleted ships' halite
+                    winner._halite += ship.halite
+
+        # Check for ship to shipyard collisions
+        for shipyard in list(board.shipyards.values()):
+            ship = shipyard.cell.ship
+            if ship is None:
+                # No collision
+                continue
+            elif ship.player_id == shipyard.player_id:
+                # Ship depositing halite
+                shipyard.player._halite += ship.halite
+                ship._halite = 0
+            else:
+                # Ship to shipyard collision
+                board._delete_shipyard(shipyard)
+                board._delete_ship(ship)
+
+        # Collect halite from cells into ships
+        for ship in board.ships.values():
+            cell = ship.cell
+            delta_halite = int(cell.halite * configuration.collect_rate)
+            if cell.shipyard_id is None and delta_halite > 0:
+                ship._halite += delta_halite
+                cell._halite -= delta_halite
+
+        # Regenerate halite in cells
+        for cell in board.cells.values():
+            next_halite = round(cell.halite * (1 + configuration.regen_rate), 3)
+            cell._halite = min(next_halite, configuration.max_cell_halite)
+
+        board._step += 1
+
+        return board
+
+
+def board_agent(agent: Callable[[Board], None]):
+    """
+    Decorator used to create an agent that modifies a board rather than an observation and a configuration
+    Automatically returns the modified board's next actions
+
+    @board_agent
+    def my_agent(board: Board) -> None:
+        ...
+    """
+    def agent_wrapper(obs, config) -> Dict[str, str]:
+        board = Board(obs, config)
+        agent(board)
+        return board.current_player.next_actions
```

Comment:
Am assuming the JSON type/domain checking is already verifying these actions are valid.

Reply:
These get filtered out and ignored in the board constructor if they're not valid